### ISSUE TITLE
custom reader questions

### ISSUE THREAD

**Issue by electronhead** [jdereg/json-io#52]
I am working on a code base that someone else wrote, so I don't have any control over how friendly the objects are to the default settings of the java-based json serialzers/deserializers out there. So far, json-io has been the easiest to get this working for the code base. I've been making progress toward json coverage for all of the objects to be serialized, and yesterday I ran into a class that did not serialize/deserialize using the default json-io settings.

This class, SpanMap, is referenced below. It "implements" Map, but doesn't do a complete job (which may be the problem). Its intention is to dynamically support dereferencing of Map values from a backing store. Perhaps the class should have not implemented Map. But serialize I must.

I wrote some JsonClass...Ex code for this class, and the writer seems to be working fine. I followed the example from your github website.

```
    writerExMap[SpanMap.class] = new JsonWriter.JsonClassWriterEx() {
      void write(Object obj, boolean showType, Writer output, Map<String, Object> args) throws IOException {
        SpanMap spanMap = (SpanMap) obj
        output.write('"totalSize":')
        output.write(String.valueOf(spanMap.getTotalSize()))
        output.write(', "spanDir":"')
        output.write(spanMap.getSpanDir())
        output.write('", "spanKey":')
        JsonWriter writer = JsonWriter.JsonClassWriterEx.Support.getWriter(args)
        writer.writeImpl(spanMap.getSpanKey(), true)
        output.write(', "map":')
        writer.writeImpl(spanMap.getMap(), true)
      }
    }
```
This JsonClassWriterEx subclass produces the following json, which is the expected output for the provided SpanMap.
```
{
  "@type": "java.util.concurrent.ConcurrentHashMap",
  "Credential": {
    "@type": "com.tlabs.datastore.index.SpanHistoryIndex",
    "map": **{
      "@type": "com.tlabs.datastore.spanmap.SpanMap",
      "totalSize": 0,
      "spanDir": "index/span/Credential/",
      "spanKey": {
        "@type": "com.tlabs.datastore.spanmap.SimpleSpanKey",
        "limit": 20000
      },
      "map": {
        "@type": "java.util.concurrent.ConcurrentSkipListMap"
      }
    }
  }
}
```
My attempt to read (deserialize) this json, however, is not working.

The code below is my attempt to instrument what's going on.
```
    readerExMap[SpanMap.class] = new JsonReader.JsonClassReaderEx() {
      Object read(Object jOb, Deque<JsonObject<String, Object>> stack, Map<String, Object> args) {
        JsonObject jObj = (JsonObject) jOb
        SpanMap spanMap = new SpanMap()
        JsonObject totalSize = jObj.get("totalSize")
        log.info("################jObj...")
        log.info(jObj.toString())
        log.info("################jObj...")
        spanMap.setTotalSize(totalSize.getPrimitiveValue())
        JsonObject spanDir = jObj.spanDir
        spanMap.setSpanDir(spanDir.getPrimitiveValue())
        JsonReader reader = JsonReader.JsonClassReaderEx.Support.getReader(args)
        JsonObject spanKeyRef = jObj.spanKey
        JsonObject spanKeyTarget = reader.getRefTarget(spanKeyRef)
        spanMap.setSpanKey(spanKeyTarget.get("spanKey"))
        JsonObject spanMapRef = jObj.spanMap
        JsonObject spanMapTarget = reader.getRefTarget(spanMapRef)
        return spanMap
      }
    }
```
The code shows that I do not understand how reading works. I tried different things as you can see and inspected values using the IntelliJ debugger. There are few code examples online, and the ones that are there have not been enough for me to get past this. Probably my problem. But still a problem.

The IntelliJ debugger shows, for example, that jObj.get("totalSize") returns null. I tried other things, but failed to access the "totalSize" field which, as you can see, is in the json string. More generally, I haven't been able to figure out how to pull values from the JsonObject without interrogating the structure shown below directly with array. I've inspected the JsonObject using IntelliJ, trying all of the documented methods. I'm definitely not understanding how reading works.

The output below is the result of logging jObj.toString(). You can see that it does not match the json listed above. For example, there is no type information for "spanKey" or "map", which are not scalar fields.
```
jObj = {com.cedarsoftware.util.io.JsonObject@1525}  size = 4
	0 = {java.util.LinkedHashMap$Entry@2862} "@keys" -> 
  		key = {java.lang.String@2614} "@keys"
  		value = {java.lang.Object[4]@2864} 
   			0 = {java.lang.String@2870} "totalSize"
   			1 = {java.lang.String@2871} "spanDir"
   			2 = {java.lang.String@2872} "spanKey"
   			3 = {java.lang.String@2873} "map"
 	1 = {java.util.LinkedHashMap$Entry@2863} "@items" -> 
  		key = {java.lang.String@2615} "@items"
  		value = {java.lang.Object[4]@2802} 
   			0 = {java.lang.Long@2803} "0"
   			1 = {java.lang.String@2804} "index/span/Credential/"
   			2 = {com.cedarsoftware.util.io.JsonObject@2805}  size = 1
    				0 = {java.util.LinkedHashMap$Entry@2832} "limit" -> "20000"
     					key = {java.lang.String@2833} "limit"
     					value = {java.lang.Long@2834} "20000"
   			3 = {com.cedarsoftware.util.io.JsonObject@2806}  size = 0
```
Questions:
1. Why does the first argument supplied to JavaClassReaderEx.read() not contain all of the information in the json string?
2. How does one access the elements in the supplied JsonObject? I thought I'd be able to access them using a Map-like accessor and cast the Object result to the desired SpanMap variable type.

This is a long post, but this is what it took to feel like I halfway described what's going on. Maybe an example would help? The bit about the JsonObject not matching the json string has me a bit worried. That said, I'm hoping to successfully deserialize in the not too distant future (using json-io).

Thanks!

-- Gary

**Comment by jdereg** [comment]
Gary,

I appreciate the detailed write up.  It looks like you are very close, and
attacking the problem class the correct way.

In your example, I see a line like below:

"map": **{

What is the '**' ?  That would throw off the JsonReader.

As for getting values from the JsonObject, fetch values from it just
as java.util.Map.  The JsonObject is extends LinkedHashMap.

Dump out the .keySet() of the read in values.  What all keys do you
see?  Dump out their associated values.

Note that the general process of reading JSON, json-io reads all of
the JSON into a Map-of-Maps representation.  Some primitive type
conversions are done in this process, but basically the JSON is held
onto in this Maps representation internally.  The next step, it
converts the Map of Maps to Java objects.  It does this using either
@type references if they exist, or Java field types (infers type from
the Java field) or it used array element type to infer the type of
objects in an array.

I know I may not have answered your question yet, but with this
information, you should be able to dig further, and then ask more
questions :-)

Best regards,

John

**Comment by electronhead** [comment]
John,

Thanks for the quick response!

The double asterisk was mistakenly left in the post to github. I was trying to format the json listing.

Here is a snippet of code…

readerExMap[SpanMap.class] = new JsonReader.JsonClassReaderEx() {
  Object read(Object jOb, Deque<JsonObject<String, Object>> stack, Map<String, Object> args) {
    JsonObject jObj = (JsonObject) jOb
    SpanMap spanMap = new SpanMap()
    log.info("################jObj...")
    log.info('jObj.get("totalSize")=' + jObj.get("totalSize"))
    jObj.keySet().each{it ->
      log.info("key=${it}")
    }
    jObj.values().each{it ->
      log.info("value=${it}")
    }
    log.info(jObj.toString())
    log.info("################jObj...")

producing in the IntelliJ console…

> 2015/10/11 12:22:27,695     main | JsonIOSerializer          INFO  - ################jObj...
> 2015/10/11 12:22:27,700     main | JsonIOSerializer          INFO  - jObj.get("totalSize")=null
> 2015/10/11 12:22:27,715     main | JsonIOSerializer          INFO  - key=@keys
> 2015/10/11 12:22:27,716     main | JsonIOSerializer          INFO  - key=@items
> 2015/10/11 12:22:27,723     main | JsonIOSerializer          INFO  - value=[totalSize, spanDir, spanKey, map]
> 2015/10/11 12:22:27,724     main | JsonIOSerializer          INFO  - value=[0, index/span/Credential/, [limit:20000], [:]]
> 2015/10/11 12:22:27,727     main | JsonIOSerializer          INFO  - [@keys:[totalSize, spanDir, spanKey, map], @items:[0, index/span/Credential/, [limit:20000], [:]]]
> 2015/10/11 12:22:27,728     main | JsonIOSerializer          INFO  - ################jObj...


Compare this output with the SpanMap section of the json string…

{"@type":"java.util.concurrent.ConcurrentHashMap","Credential":{"@type":"com.tlabs.datastore.index.SpanHistoryIndex","map":{"@type":"com.tlabs.datastore.spanmap.SpanMap","totalSize":0, "spanDir":"index/span/Credential/", "spanKey":{"@type":"com.tlabs.datastore.spanmap.SimpleSpanKey","limit":20000}, "map":{"@type":"java.util.concurrent.ConcurrentSkipListMap"}}}}

I didn’t know that I would need to access the application objects through indirection (@keys,@items), but that is easy enough to do. What I don’t understand is why the “spanKey" and “map” values are not typed as a SimpleSpanKey and ConcurrentSkipListMap, respectively. I assumed that json-io would use the @type information to build the application objects and put them in the @items list. One issue might be that ConcurrentSkipListMap is a Map, so that might need special handling, but SimpleSpanKey is not a Map. How would I find the proper classes to instantiate objects from the execution context of the “read” method?

My basic assumption here is that creating a custom serializer for a particular class does not require the creation of custom serializers for all instance variables in the particular class. In this example, I assumed that I didn’t need to create a custom serializer for SpanKey because it was in a class that required a custom serializer. Was this assumption correct?

Thanks for your help.

**Comment by jdereg** [comment]
​You mention that you would expect the 'spanlKey' and 'map' values ​tp be
typed as SimpleSpanKey and ConcurrentSkipListMap, but when I look at the
JSON example you included, I see
{"@type":"com.tladbs.datastore.spanmap.SimpleSpanKey" and I see map's type
as "java.util.concurrent.ConcurrentSkipListMap"} which is exactly the type
of those classes.  The package name *is* part of the class name.  Were you
expect json-io to drop the package portion of the class name?

**Comment by electronhead** [comment]
Yes, the json looks great (thank-you!). I didn’t include the whole class path. I abbreviated. Sorry.

I’ve made some progress, I think, mostly because I realized that IntelliJ was hiding most object state by default. Operator ignorance. Apologies.

So the SimpleSpanKey element of @items has a type as below...

jObj[“@items”][2].getType() == “com.tlabs.datastore.spanmap.SimpleSpanKey”

To the point, here is the current iteration of the “read” code for SpanMap...

  readerExMap[SpanMap.class] = new JsonReader.JsonClassReaderEx() {
    Object read(Object jOb, Deque<JsonObject<String, Object>> stack, Map<String, Object> args) {
      JsonObject jObj = (JsonObject) jOb
      SpanMap spanMap = (SpanMap) jObj.getTarget()
      [jObj["@keys"], jObj["@items"]].transpose().collect { it ->
        String key = (String) it[0]
        Object item = it[1]
        spanMap[key] = deserial(item, args)
      }
      return spanMap
    }
  }
}
Object deserial(Object item, args) {
  if(item instanceof JsonObject) {
    JsonReader reader = JsonReader.JsonClassReaderEx.Support.getReader(args)
    return reader.jsonObjectsToJava((JsonObject) item)
  } else { // already deserialized
    return item
  }
}
Knowing what was there helped tremendously. Made an educated guess about the "deserial" method’s use of jsonObjectsToJava, which in hindsight seems obvious.

On to other serialization issues. But I think I have the basic idea of json-io working for me.

Thanks for your help.

— Gary
